src/payment/data/sagas.js

```diff
@@ -310,43 +311,78 @@ export function* handleSubmitPayment({ payload }) {
-        ? null : state.payment.basket.payments[0].paymentNumber));
-
-      if (!basketId || !paymentNumber) {
-        // This shouldn't happen.
-        //   I don't think we need to banner... shouldn't our parent calls recover? (They invoke this)
-        keepPolling = false;
-        yield put(pollPaymentState.fulfill());
-        return;
-      }
+  const delaySecs = getConfig().PAYMENT_STATE_POLLING_DELAY_SECS || DEFAULT_PAYMENT_STATE_POLLING_DELAY_SECS;
+
+  const [basketId, paymentNumber, paymentState] = yield select(state => ([
+    /* basketId */
+    state.payment.basket.basketId,
+    /* paymentNumber */
+    (state.payment.basket.payments.length === 0
+      ? null : state.payment.basket.payments[0].paymentNumber),
+    /* paymentState */
+    state.payment.basket.paymentState,
+  ]));
+
+  if (!POLLING_PAYMENT_STATES.includes(paymentState)) {
+    yield put(pollPaymentState.fulfill());
+    return;
+  }
+
+  try {
+    while (true) { //  o/ o/ ive got to break free! o/ o/ ... or throw.
+      try {
+        if (!basketId || !paymentNumber) {
+          // noinspection ExceptionCaughtLocallyJS
+          throw new ReferenceError('Invalid Basket Id or Payment Number');
+        }
 
-      const result = yield call(PaymentApiService.getCurrentPaymentState, paymentNumber, basketId);
-      yield put(paymentStateDataReceived(result));
+        const result = yield call(PaymentApiService.getCurrentPaymentState, paymentNumber, basketId);
+
+        yield put(pollPaymentState.received(result));
+
+        if (!(yield select(state => state.payment.basket.paymentStatePolling.keepPolling))) {
+          yield put(pollPaymentState.fulfill());
+          break;
+        } else {
+          yield delay(SECS_AS_MS(delaySecs));
+        }
+      } catch (innerError) {
+        const shouldExitExecution = innerError instanceof ReferenceError;
+
+        logError(innerError, { basketId, paymentNumber, idFatal: shouldExitExecution });
+
+        if (shouldExitExecution) {
+          // noinspection ExceptionCaughtLocallyJS
+          throw innerError;
+        }
+
+        yield put(pollPaymentState.received({ state: PAYMENT_STATE.HTTP_ERROR }));
+
+        const currentErrorCount = yield select(state => state.payment.basket.paymentStatePolling.errorCount);
+
+        if (currentErrorCount === 0) {
+          // noinspection ExceptionCaughtLocallyJS
+          throw innerError;
+        }
 
-      if (!(yield select(state => state.payment.basket.paymentStatePolling.keepPolling))) {
-        keepPolling = false;
-        yield put(pollPaymentState.fulfill());
-      } else {
         yield delay(SECS_AS_MS(delaySecs));
       }
-    } catch (error) {
-      // We dont quit on error.
-      // yield call(handleErrors, error, true);
-      yield put(paymentStateDataReceived({ state: PAYMENT_STATE.HTTP_ERROR }));
-      yield delay(SECS_AS_MS(delaySecs));
     }
+  } catch (error) {
+    const basketMessageError = generateApiError([
```

Comment:
Runtime Value Errors as well as all others if they were thrown from our inner loop need to be wrapped so they prompt the user the reload the page because their basket state is unexpected.